src/mscorlib/src/System/Text/StringBuilder.cs

```diff
@@ -999,6 +999,130 @@ void ISerializable.GetObjectData(SerializationInfo info, StreamingContext contex
             return this;
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            Contract.Ensures(Contract.Result<StringBuilder>() != null);
+
+            if (value != null && value.Length > 0)
+            {
+                if (value != this)
```

Comment:
@AlexRadch I agree -- just calling `sb2.ToString()` will not prevent the race. 

However, once you have the string produced by calling `ToString()` you can safely pass that in to `Append` since that is now an immutable object.

What I want to make sure if that this API (which introduces a way to pass in a *mutable* object in) is resilient to changes to the *mutable* object.

If you look at `StringBuilder` there are quite a few places where we do thread safety checks. Those checks are mostly there to make sure that the internal state of the `StringBuilder` is not corrupted by 'bad' input.